File: frontend/src/lib/name-range.ts
/**
 * One `[a-b]` range in a component name expands to that many components, so
 * "Disk[1-5]" creates Disk1 … Disk5 in a single submit. Any other text is
 * left alone, which keeps a literal name like "Disk[1-5]" impossible but
 * every ordinary name unaffected.
 *
 * Shared by the device-type template dialog and every device-page component
 * dialog, so multi-add means the same thing everywhere.
 */
export const NAME_RANGE_RE = /\[(\d+)-(\d+)\]/

/** Refuse to fan out beyond this in one submit - a typo like [1-99999]
 * shouldn't try to create 99k rows. */
export const RANGE_CAP = 128

/**
 * "Disk[1-5]" → ["Disk1", …, "Disk5"]. Returns `[name]` unchanged when there
 * is no range, the bounds are reversed or unparseable, or the span exceeds
 * {@link RANGE_CAP} - callers can then treat it as a plain single name.
 */
export function expandNameRange(name: string): string[] {
  const m = name.match(NAME_RANGE_RE)
  if (!m) return [name]
  const lo = Number(m[1])
  const hi = Number(m[2])
  if (!Number.isFinite(lo) || !Number.isFinite(hi) || hi < lo) return [name]
  if (hi - lo + 1 > RANGE_CAP) return [name]
  const out: string[] = []
  for (let i = lo; i <= hi; i++)
    out.push(name.replace(NAME_RANGE_RE, String(i)))
  return out
}

/** True when the name carries a range that {@link expandNameRange} will act
 * on - for showing a "creates N" hint before the user submits. */
export function hasNameRange(name: string): boolean {
  return expandNameRange(name).length > 1
}

/**
 * Run one create per expanded name, in order. Sequential on purpose: the
 * server enforces name uniqueness per device, so a clash partway through a
 * range reports the name that clashed instead of an unordered pile of
 * rejections. Whatever was created before the failure stays created.
 */
export async function createEach<T>(
  names: string[],
  post: (name: string) => Promise<T>
): Promise<{ last: T; count: number }> {
  let last!: T
  for (const name of names) last = await post(name)
  return { last, count: names.length }
}

File: api/name_range.py
"""One ``[a-b]`` range in a component name fans out to that many components.

The expansion used to live only in the frontend dialogs, so a template created
through the raw API with "RF[1-2]" became one literally-named row (#111 test
seeding surfaced it). The server is the right place for a naming contract:
now every create path - dialog, script, import - means the same thing.

Mirrors ``frontend/src/lib/name-range.ts`` exactly: one range per name, bounds
must be ordered, spans over the cap fall back to a plain single name.
"""
from __future__ import annotations

import re

NAME_RANGE_RE = re.compile(r"\[(\d+)-(\d+)\]")

#: Refuse to fan out beyond this in one create - a typo like [1-99999]
#: must not try to make 99k rows.
RANGE_CAP = 128


def expand_name_range(name: str) -> list[str]:
    """"Disk[1-5]" → ["Disk1", …, "Disk5"]; anything else → [name]."""
    m = NAME_RANGE_RE.search(name or "")
    if not m:
        return [name]
    lo, hi = int(m.group(1)), int(m.group(2))
    if hi < lo or hi - lo + 1 > RANGE_CAP:
        return [name]
    return [NAME_RANGE_RE.sub(str(i), name, count=1) for i in range(lo, hi + 1)]


def range_error(name: str) -> str | None:
    """Why a name's range shorthand can't be honoured, or ``None`` when the
    name is a plain name or a usable ``[a-b]`` range. Serializers surface this
    so a typo is refused instead of stored literally."""
    if re.search(r"\{\d+-\d+\}", name):
        return "Ranges use square brackets: [1-24]."
    if len(NAME_RANGE_RE.findall(name)) > 1:
        return "Only one [a-b] range per name."
    if NAME_RANGE_RE.search(name) and len(expand_name_range(name)) == 1:
        return f"A range must count up and cover at most {RANGE_CAP} names."
    return None
